src/lib.rs

```diff
@@ -1,14 +1,186 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+use embedded_hal::blocking::i2c::{Write, WriteRead};
+
+const AK09915_ADDRESS: u8 = 0x0C;
+
+// Register Addresses
+#[allow(dead_code)]
+enum Register {
+    WIA1 = 0x00,   // Device ID Register 1
+    WIA2 = 0x01,   // Device ID Register 2
+    ST1 = 0x10,    // Data Status Register 1
+    HXL = 0x11,    // Data Register - X-axis Magnetic Data Low Byte
+    HXH = 0x12,    // Data Register - X-axis Magnetic Data High Byte
+    HYL = 0x13,    // Data Register - Y-axis Magnetic Data Low Byte
+    HYH = 0x14,    // Data Register - Y-axis Magnetic Data High Byte
+    HZL = 0x15,    // Data Register - Z-axis Magnetic Data Low Byte
+    HZH = 0x16,    // Data Register - Z-axis Magnetic Data High Byte
+    TMPS = 0x17,   // Temperature Sensor Data Register
+    ST2 = 0x18,    // Data Status Register 2
+    CNTL2 = 0x31,  // Control Register 2
+    CNTL3 = 0x32,  // Control Register 3
+    TS1 = 0x33,    // Self Test Register 1
+    TS2 = 0x34,    // Self Test Register 2
+    I2CDIS = 0x3A, // I2C Disable Register
+}
+
+// AK09915 Mode Settings - Corresponding to Control Register 2
+#[allow(dead_code)]
+pub enum Mode {
+    PowerDown = 0x00,
+    Single = 0x01,
+    Cont10Hz = 0x02,
+    Cont20Hz = 0x04,
+    Cont50Hz = 0x06,
+    Cont100Hz = 0x08,
+    Cont200Hz = 0x0A,
+    Cont1Hz = 0x0C,
+    SelfTest = 0x10,
+}
+
+impl From<Mode> for u8 {
+    fn from(mode: Mode) -> Self {
+        mode as u8
+    }
+}
+
+impl From<Register> for u8 {
+    fn from(register: Register) -> Self {
+        register as u8
+    }
+}
+
+pub struct Ak09915<I2C> {
+    pub i2c: I2C,
+    pub address: u8,
+}
+
+impl<I2C, E> Ak09915<I2C>
+where
+    I2C: Write<Error = E> + WriteRead<Error = E>,
+{
+    pub fn new(i2c: I2C) -> Self {
+        Self {
+            i2c,
+            address: AK09915_ADDRESS,
+        }
+    }
+
+    fn write_register(&mut self, register: u8, value: u8) -> Result<(), E> {
+        self.i2c.write(self.address, &[register, value])
+    }
+
+    fn read_register(&mut self, register: u8) -> Result<u8, E> {
+        let mut buffer = [0u8];
+        self.i2c
+            .write_read(self.address, &[register], &mut buffer)
+            .and(Ok(buffer[0]))
+    }
+
+    pub fn init(&mut self) -> Result<(), E> {
+        // Soft reset device and put on continuous measurement
+        self.reset()?;
+        self.set_mode(Mode::PowerDown)?;
+        Ok(())
+    }
+
+    pub fn reset(&mut self) -> Result<(), E> {
+        // Soft reset device
```

Comment:
Hint: use [doc comments syntax](https://doc.rust-lang.org/rust-by-example/meta/doc.html) here and everywhere else :)
```suggestion

    /// Soft reset device
    pub fn reset(&mut self) -> Result<(), E> {
```

Reply:
actually, might be better to do it in another PR, if/when we decide to prepare this to be a published crate to crates.io, then we just follow this: https://doc.rust-lang.org/book/ch14-02-publishing-to-crates-io.html#publishing-a-crate-to-cratesio